library/Fission/Error.hs

```diff
@@ -25,3 +27,18 @@ fromMaybe err okHandler = maybe (openLeft err) (Right . okHandler)
 
 fromMaybe' :: IsMember err errs => err -> Maybe a -> Either (OpenUnion errs) a
 fromMaybe' err = fromMaybe err identity
+
+retryOnErr ::
+  Monad m
+  => (a -> m Bool)
+  -> Natural
+  -> m a
+  -> m a
+retryOnErr check times action = do
+  result <- action
+  check result >>= \case
+    True -> return result
+    False -> 
+      if times > 0
+        then retryOnErr check (times - 1) action
+        else return result
```

Comment:
Two things:

# 1. Formatting

Let's try to keep the formatting consistent:

```haskell
  result <- action
  check result >>= \case
    True -> 
      return result

    False -> 
      if times > 0
        then retryOnErr check (times - 1) action
        else return result
```

# 2. Multiple Heads

The current version is pretty imperative. The idiomatic way to write this is:

```haskell
retryOnErr _     0          action = action
retryOnErr check retryCount action =
  action >>= check >>= \case
    True  -> return result
    False -> retryOnErr check (times - 1) action
```

or

```haskell
retryOnErr _     0          action = action
retryOnErr check retryCount action =
  ok <- check <$> action
  if ok
    then return result
    else retryOnErr check (times - 1) action
```

Reply:
oh yeah good point on the base case^^

I think I still need to get `result` out of `action` however. So:
```haskell
retryOnErr check times action = do
  result <- action
  check result >>= \case
    True -> return result
    False -> retryOnErr check (times - 1) action
```
altho it sounds like I should avoid blending `do` notation & `bind` notation
so maybe:
```haskell
retryOnErr _ 0 action = action
retryOnErr check times action = do
  result <- action
  ok <- check result
  if ok
    then return result
    else retryOnErr check (times - 1) action
````

Reply:
```haskell
result <- action
check result >>= \case

-- is just sugar for

action >>= check >>= \case
```

Reply:
I'm fine with this version of yours:

```haskell
retryOnErr _ 0 action = action
retryOnErr check times action = do
  result <- action
  ok <- check result
  if ok
    then return result
    else retryOnErr check (times - 1) action
```

You're right about the types in my example with `<$>`. We could write it this way:

```haskell
do
  ok <- check =<< action
  if ok
    then return result
    else retryOnErr check (times - 1) action
```

or even this way if you're going to reuse that a lot in the body (probably not this case)

```haskell
do
  ok <- checkM action
  if ok
    then return result
    else retryOnErr check (times - 1) action
where
  checkM foo = check =<< foo
```

Reply:
These are all the same to the compiler. Which reads best to you?

Reply:
but in the `if` statement  you need to return `result` so you can't just bind it through to `check`

I think our options are:
```haskell
retryOnErr _ 0 action = action
retryOnErr check times action = do
  result <- action
  ok <- check result
  if ok
    then return result
    else retryOnErr check (times - 1) action
```
or
```haskell
retryOnErr _ 0 action = action
retryOnErr check times action = 
  action >>= \result ->
    check result >>= \case
      True -> return result
      False -> retryOnErr check (times - 1) action
```
if we want bind syntax

Reply:
Oooh, right we need to return `result`! Right right. I'd use the `do` version 👍